unit/deserialize.go

```diff
@@ -43,36 +43,83 @@ var (
 	ErrLineTooLong = fmt.Errorf("line too long (max %d bytes)", SYSTEMD_LINE_MAX)
 )
 
-// Deserialize parses a systemd unit file into a list of UnitOption objects.
+// Deserialize parses a systemd unit file into a list of UnitOptions.
 func Deserialize(f io.Reader) (opts []*UnitOption, err error) {
-	lexer, optchan, errchan := newLexer(f)
+	_, options, err := DeserializeAll(f)
+	return options, err
+}
+
+// DeserializeSections deserializes into a list of UnitSections.
+func DeserializeSections(f io.Reader) ([]*UnitSection, error) {
+	sections, _, err := DeserializeAll(f)
+	return sections, err
+}
+
+// DeserializeAll deserializes into UnitSections and UnitOptions.
+func DeserializeAll(f io.Reader) ([]*UnitSection, []*UnitOption, error) {
+
+	lexer, secchan, optchan, errchan := newLexer(f)
+
 	go lexer.lex()
 
-	for opt := range optchan {
-		opts = append(opts, &(*opt))
+	sections := []*UnitSection{}
+	options := []*UnitOption{}
+
+	var done bool
+
+	for {
+		select {
+		case opt, ok := <-optchan:
+			if !ok {
+				done = true
+			} else {
+				// add to options
+				options = append(options, &(*opt))
+
+				// add to newest section entries.
+				s := len(sections) - 1
```

Comment:
I fear this will panic if we manager to get here with `len == 0`.

Reply:
For that to happen the optchan would need to fire before the secchan. I don't believe that's possible given the code flow. All optchans fires are down code stream from secchan fires. 

I can certainly add a defensive check there though if you want.

Reply:
Just for clarity, it doesn't matter the order in which they **fire**. What matters is the order in which `select` **receives** from them. As those are buffered channels, both can be ready to poll at the same time, in which case reception will be in random order (across the two channels).

FIFO order is guaranteed only on the single channel, i.e. each one separately.